File: Sources/Services/BackgroundMusicPlayer.swift
import AVFAudio
import Foundation

/// Owns the single foreground background-music player for the iOS app,
/// mirroring the Android `BackgroundMusicPlayer` behavior: looping track,
/// lifecycle pause/resume, and lower volume in Power Save Mode.
@MainActor
final class BackgroundMusicPlayer {
    private var player: AVAudioPlayer?
    var isEnabled = true {
        didSet {
            guard !isEnabled else { return }
            player?.pause()
        }
    }

    func play() {
        guard isEnabled else { return }
        let audioPlayer = player ?? createPlayer()
        guard let audioPlayer else { return }
        player = audioPlayer
        refreshVolume()
        if !audioPlayer.isPlaying {
            audioPlayer.play()
        }
    }

    func pause() {
        player?.pause()
    }

    private func refreshVolume() {
        let volume = ProcessInfo.processInfo.isLowPowerModeEnabled ? Self.lowPowerVolume : Self.normalVolume
        player?.volume = volume
    }

    private func createPlayer() -> AVAudioPlayer? {
        guard let url = Bundle.main.url(forResource: "collapse_arcade_vibe", withExtension: "mp3") else {
            return nil
        }
        let session = AVAudioSession.sharedInstance()
        try? session.setCategory(.ambient, options: [.mixWithOthers])
        try? session.setActive(true)
        guard let audioPlayer = try? AVAudioPlayer(contentsOf: url) else { return nil }
        audioPlayer.numberOfLoops = -1
        return audioPlayer
    }

    private static let normalVolume: Float = 0.18
    private static let lowPowerVolume: Float = 0.10
}

File: android/app/src/main/java/com/collapse/game/services/BackgroundMusicPlayer.kt
package com.collapse.game.services

import android.content.Context
import android.media.MediaPlayer
import android.os.PowerManager
import com.collapse.game.R

/** Owns the single foreground background-music player for the Android app. */
class BackgroundMusicPlayer(context: Context) : AutoCloseable {
    private val appContext = context.applicationContext
    private val powerManager = appContext.getSystemService(PowerManager::class.java)
    private var player: MediaPlayer? = null

    fun play() {
        val mediaPlayer = player ?: createPlayer()?.also { player = it } ?: return
        refreshVolume()
        if (!mediaPlayer.isPlaying) mediaPlayer.start()
    }

    fun pause() {
        val mediaPlayer = player ?: return
        if (mediaPlayer.isPlaying) mediaPlayer.pause()
    }

    fun refreshVolume() {
        val volume = if (powerManager?.isPowerSaveMode == true) LOW_POWER_VOLUME else NORMAL_VOLUME
        player?.setVolume(volume, volume)
    }

    override fun close() {
        player?.release()
        player = null
    }

    private fun createPlayer(): MediaPlayer? = MediaPlayer.create(appContext, R.raw.collapse_arcade_vibe)?.apply {
        isLooping = true
    }


    private companion object {
        const val NORMAL_VOLUME = 0.18f
        const val LOW_POWER_VOLUME = 0.10f
    }
}
